src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,612 @@
+        }
+        column_pruner * found = nullptr;
+        if (tree_stack.back() != nullptr) {
+          // tree_stack can have a nullptr in it if the scheam we are looking through
+          // has an entry that does not match the tree
+          auto found_it = tree_stack.back()->children.find(name);
+          if (found_it != tree_stack.back()->children.end()) {
+            found = &(found_it->second);
+            int parent_mapped_schema_index = tree_stack.back()->s_id;
+            num_children_map[parent_mapped_schema_index]++;
+          }
+        }
+
+        if (schema_item.__isset.type) {
+          if (found != nullptr) {
+            int mapped_chunk_index = found->c_id;
+            int mapped_schema_index = found->s_id;
+            chunk_map[mapped_chunk_index] = chunk_index;
+            schema_map[mapped_schema_index] = schema_index;
+            num_children_map[mapped_schema_index] = 0;
+          }
+          // this is a leaf node
+          chunk_index++;
+        } else {
+          // a non-leaf node
+          if (found != nullptr) {
+            int mapped_schema_index = found->s_id;
+            schema_map[mapped_schema_index] = schema_index;
+            num_children_map[mapped_schema_index] = 0;
+          }
+        }
+
+        if (num_children > 0) {
+          tree_stack.push_back(found);
+          num_children_stack.push_back(num_children);
+        } else {
+          // go back up the stack/tree removing children until we hit one with more children
+          bool done = false;
+          while (!done) {
+            int parent_children_left = num_children_stack.back() - 1;
+            if (parent_children_left > 0) {
+              num_children_stack.back() = parent_children_left;
+              done = true;
+            } else {
+              tree_stack.pop_back();
+              num_children_stack.pop_back();
+            }
+ 
+            if (tree_stack.size() == 0) {
+              done = true;
+            }
+          }
+        }
+      }
+
+      // If there is a column that is missing from this file we need to compress the gather maps
+      //  so there are no gaps
+      std::vector<int> final_schema_map;
+      final_schema_map.reserve(schema_map.size());
+      for (auto it = schema_map.begin(); it != schema_map.end(); it++) {
+        final_schema_map.push_back(it->second);
+      }
+
+      std::vector<int> final_num_children_map;
+      final_num_children_map.reserve(num_children_map.size());
+      for (auto it = num_children_map.begin(); it != num_children_map.end(); it++) {
+        final_num_children_map.push_back(it->second);
+      }
+
+      std::vector<int> final_chunk_map;
+      final_chunk_map.reserve(chunk_map.size());
+      for (auto it = chunk_map.begin(); it != chunk_map.end(); it++) {
+        final_chunk_map.push_back(it->second);
+      }
+
+      return column_pruning_maps{final_schema_map, final_num_children_map, final_chunk_map};
```

Comment:
I thought that the C++ code was smart enough to do the move automatically. I could be wrong but on the CUDF code I keep tell people not to put the std::move on a return, but because we are putting it into a struct I am not 100% sure on it.

Reply:
I wrote sample code with my own struct with operator= and the copy ctor, then I had a container struct like yours, but with std::vector<my_struct>. Without std::move there was an extra copy when constructing the return value. There wasn't an extra copy outbound of the function (so the return-value-optimization worked, e.g the return by value was optimized out).